alphagov/paas-cf
# [#157693388] London for real

## What

Allow to deploy Cloud Foundry in dev for both eu-west-1 and eu-west-2. This will allow us to choose where we want to deploy our dev environments.

The region is controlled by the value of `AWS_DEFAULT_REGION`.

To make life easier we pass all feature flag env vars (`ENABLE_DATADOG`, `NEW_ACCOUNT_EMAIL_ADDRESS`) to pipelines.sh/self-update-pipeline.sh in Concourse which means you don't have to set SELF_UPDATE_PIPELINE=false anymore if you enable these flags. This means you can always test the self update pipeline functionality too and you don't have to update the pipeline manually after every commit. (If you're pushing local changes then you still need SELF_UPDATE_PIPELINE=false)

As there are no m3.* or c3.* instance types in eu-west-2 we had to change some instance types.

## Instance type changes

| VM                      | From         | To          |
| ----------------------- | ------------ | ----------- |
| compilation (Concourse) | c3.large     | c4.large    |
| compilation (paas-cf)   | c3.large     | c4.large    |
| router                  | c3.large     | c4.large    |
| CF DB                   | db.m3.medium | db.m4.large |
| medium                  | m3.medium    | m4.large    | 

The CF DB CPU usage is constantly between 60-90% in prod, therefore we can only move to an m4.large. This could mean a couple of minutes of downtime if changed.

## Original m3.medium instances

| VM        | Count | CPU usage | Suggested  | Cost difference |
| --------- | ----- | --------- | ---------- | --------------- |
| cc-worker | 2     | ~5%       | t2.medium  | - 2 * 0.021     |
| diego-api | 2     | <25%      | t2.medium  | - 2 * 0.021     |
| doppler   | 6     | ~50%      | m4.large   | + 6 * 0.043     |
| ingestor  | 2     | 10-20%    | t2.medium  | - 2 * 0.021     |
| log-api   | 2     | 20-60%    | m4.large   | + 2 * 0.043     |
| nats:     | 2     | ~5%       | t2.medium  | - 2 * 0.021     |
| scheduler | 2     | <20%      | t2.medium  | - 2 * 0.021     |
| uaa       | 2     | ~10%      | t2.medium  | - 2 * 0.021     |
| SUM       | 20    | -         | -          | +$0.092/hour    |

The queue instances are m3.medium only in prod with a ~5% cpu usage, so we'll change them to t2.medium (small cost saving not counted here)

m3.medium -> m4.large price difference: 0.116 - 0.073 = +0.043 per hour
m3.medium -> t2.medium price difference: 0.073 - 0.052 = -0.021 per hour

*Note: m4.large and t2.medium pricing is for London*

Additional changes:
 - When we trigger the app and api availability tests we upload a job file to S3. This was done in one task and we always installed awscli on the top of the `cf-acceptance-tests` Docker image. Now this is in three separate tasks.

## How to review

1. Create, update and delete a Cloud Foundry environment in both regions:

```
AWS_DEFAULT_REGION=... DEPLOY_ENV=... BRANCH=london_for_real_157693388 ENABLE_DATADOG=true make dev pipelines
```

1. Check that all tasks in Concourse runs successfully. Check credentials rotation and cert rotation (only leafs).

1. Test the create-user script.

## How to release

❗️ There is a temporary commit to set the branch for kick-off/destroy tasks - this needs to be removed

Please review and merge https://github.com/alphagov/paas-aws-account-wide-terraform/pull/126 first.

Talk with @alext or @henrytk how to release this change to prod as the CF database instance type will be changed.

## Who can review

Not @henrytk or me.
